vllm/model_executor/models/utils.py

```diff
@@ -437,60 +519,25 @@ def set_cpu_offload_max_bytes(max_bytes: int) -> None:
     _CPU_OFFLOAD_MAX_BYTES = max_bytes
 
 
-def maybe_offload_to_cpu(module: torch.nn.Module) -> torch.nn.Module:
-    device = next(module.parameters()).device
-
-    if device == torch.device("cpu"):
-        return module
-
-    global _CPU_OFFLOAD_MAX_BYTES, _CPU_OFFLOAD_BYTES
-    if _CPU_OFFLOAD_BYTES >= _CPU_OFFLOAD_MAX_BYTES:
-        return module
-
-    pin_memory = is_pin_memory_available()
-
-    # offload parameters to CPU
-    # use pin_memory if possible, which helps cudagraph capture speed
-    offloaded_parameters = False
-    for p in module.parameters():
-        if _CPU_OFFLOAD_BYTES >= _CPU_OFFLOAD_MAX_BYTES:
-            # we use per-parameter offloading
-            # one module might have some parameters offloaded and some not
-            break
-
-        # `torch.empty_like` does not support `pin_memory` argument
-        cpu_data = torch.empty_strided(size=p.data.size(),
-                                       stride=p.data.stride(),
-                                       dtype=p.data.dtype,
-                                       layout=p.data.layout,
-                                       device='cpu',
-                                       pin_memory=pin_memory)
-        cpu_data.copy_(p.data)
-        p.data = cpu_data
-        _CPU_OFFLOAD_BYTES += p.data.numel() * p.data.element_size()
-        offloaded_parameters = True
-
-    if offloaded_parameters:
-        original_forward = module.forward
-
-        def forward(*args, **kwargs):
-            module.forward = original_forward
-            device_state = {
-                # here we blindly call `to(device)`
-                # if the parameter is already on the device, it will be a no-op
-                k: v.to(device, non_blocking=True)
-                for k, v in module.state_dict().items()
-            }
-            output = functional_call(module,
-                                     device_state,
-                                     args=args,
-                                     kwargs=kwargs)
-            module.forward = forward
-            return output
+@contextmanager
+def maybe_offload_to_cpu():
+    old_empty = torch.empty
 
-        module.forward = forward
+    def fake_empty(*args, **kwargs):
+        global _CPU_OFFLOAD_MAX_BYTES, _CPU_OFFLOAD_BYTES
+        tensor = old_empty(*args, **kwargs)
+        if tensor.device != torch.device("cpu") and \
+            _CPU_OFFLOAD_BYTES < _CPU_OFFLOAD_MAX_BYTES:
+            _CPU_OFFLOAD_BYTES += tensor.numel() * tensor.element_size()
+            offloaded_tensor = OffloadedTensor(tensor)
+            return offloaded_tensor
+        return tensor
 
-    return module
+    torch.empty = fake_empty
```

Comment:
One thing to call out is that the monkey patching here will allow you to override `torch.empty`, but not any `at::empty()` calls that come from C++ anywhere in the dispatcher. I'm not sure if the particular code you're running is actually running into this, but the way we normally handle "factory functions that you want to override to return tensor subclasses" is with a `TorchDispatchMode`:
```
from torch.utils._python_dispatch import TorchDispatchMode

class OffloadedTensorMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        rs = func(*args, **kwargs)
        if func is torch.ops.aten.empty.memory_format and rs.device != "cpu" and ...:
            rs = OffloadedTensor(rs)
        return rs
```

Reply:
this is very helpful!